DRAFT: Update wasmer version

ByteNacked opened the pull request:
Resolves # .

-
-
-

@reviewer-or-team

breathx commented on sandbox/host/src/sandbox.rs:
```diff
@@ -243,11 +254,55 @@ impl SandboxInstance {
+        &self,
+        name: &str,
+        value: Value,
+    ) -> std::result::Result<Option<()>, error::Error> {
```
```suggestion
    ) -> Result<Option<()>, error::Error> {
```

breathx commented on sandbox/host/src/sandbox/wasmer_backend.rs:
```diff
@@ -309,43 +325,64 @@ pub fn instantiate(
+    let mut import_object = sandbox_wasmer::Imports::new();
+    import_object.register_namespace("env", exports);
+
+    // This is sound since we're only instantiating the module and `start` function is restricted by our code checks.
```
Do you mean that there's nothing we actually run so we don't need any context because nothing will access this global value etc?

ByteNacked replied:
Yes, that's correct. No code can access globals or call syscalls, as the start function is not allowed by Code checks. Therefore, this context loads unnecessarily.

It seems both of these can be removed at module instantiation:
```
let instance = SandboxContextStore::using(sandbox_context, || {
        dispatch_function_env::using(&mut dispatch_func_env, || {
```

breathx replied:
Could you extend comment a bit to avoid future misunderstanding please?

ByteNacked replied:
Extended comment

ByteNacked commented on sandbox/host/src/sandbox/wasmer_backend.rs:
```diff
@@ -20,28 +20,59 @@
 
 environmental::environmental!(SandboxContextStore: trait SandboxContext);
 
+// Hack to allow mutable `environmental!` definition per module
```
```suggestion
// Hack to allow multiple `environmental!` definition per module
```

ByteNacked commented:
It seems the current approach doesn't really work. During syscalls, second mutable borrow errors happen.

NikVolf commented:
Have you resynced the whole Vara history with it @ByteNacked ?

ByteNacked commented:
> Have you resynced the whole Vara history with it @ByteNacked ?

Node re-syncing for vara, vara-dev is in progress ATM.

ByteNacked commented:
> Update2: The problem with random segfaults was that the Module in the LRU cache outlived the Store. I fixed it.

A bit more detail about the segfaults that initially appeared after upgrading Wasmer to version 4.

This looks like an API vulnerability in Wasmer (it appears only in version 4 due to architectural changes in Wasmer) because the segfaults were caused by the Module we stored in the LRU cache containing references to a memory buffer with compiled code stored in the Store. Thus, when the Store was dropped (see [Store::clear](https://github.com/gear-tech/gear/blob/8ff08219798050aa2576b6f1d7eb2f5cd7ab017b/sandbox/host/src/sandbox.rs#L540)), the reference to this buffer outlived the Store and became a dangling pointer.

The problem was resolved by serializing the module before writing it to the LRU cache, similar to what we do when writing to the FS cache. Inside the module, there is an `Artifact` object that can be either `Archived` or `Plain`. If it is `Plain`, it can contain a reference to the compiled code in the `Store`. If it is `Archived`, it is self-contained. During serialization, it transforms into `Archived` and then gets serialized. There is no API to simply make the module `Archived` (there is, but it's low-level and very fragile according to comments in the code).

cc @grishasobol

ByteNacked commented:
Waiting for resync of the whole Vara history :gear:

ByteNacked commented:
@breathx Please re-review due to substantial changes.

ByteNacked commented:
:exclamation: :exclamation: :exclamation:  One more problem was found. Both syncs with the testnet and mainnet were successful, but during the testnet sync, the node stucks on a certain block (every sync attempt). After restarting the node, the sync continues as normal.

ByteNacked commented:
Waiting for one more review/approval @grishasobol  @gshep

ByteNacked commented:
Update:

* After discussion renamed some entities in sandbox-host crate
* Refactored/simplified code a little bit, move access to `gas_global` into `FunctionEnv`
* Added same store check in `StoreRefCell`

gshep commented on Cargo.toml:
```diff
@@ -173,12 +174,12 @@ syn = "2.0.71"
-url = "2.5.2"
-wat = "1.0.83"
+url = "2.5.0"
+wat = "1.0.71"
```
do we need to downgrade these two?

gshep commented on sandbox/host/src/sandbox.rs:
```diff
@@ -243,11 +251,55 @@ impl SandboxInstance {
+
+    /// Get the value from a global with the given `name`. Only for usage in signal handler.
+    ///
+    /// Returns `Some(_)` if the global could be found.
```
```suggestion
    /// Returns `Some(_)` if the global has been found.
```

gshep commented on sandbox/host/src/sandbox.rs:
```diff
@@ -243,11 +251,55 @@ impl SandboxInstance {
-            BackendInstance::Wasmer(wasmer_instance) => {
-                wasmer_set_global(wasmer_instance, name, value)
+            BackendInstanceBundle::Wasmer { instance, store } => unsafe {
+                // We cannot use `store.borrow_mut()` in single handler context because it's already borrowed during `invoke` call.
```
```suggestion
                // We cannot use `store.borrow_mut()` in signal handler context because it's already borrowed during `invoke` call.
```

gshep replied:
The same in the next func too

gshep commented on sandbox/host/src/sandbox/wasmer_backend/store_refcell.rs:
```diff
@@ -0,0 +1,343 @@
+pub struct BorrowScopeError;
+
+impl StoreRefCell {
+    /// Create new `StoreRefCell` with provided `Store
```
```suggestion
    /// Create new `StoreRefCell` with provided `Store`
```

gshep reviewed (approved):
Awesome job! :100:

grishasobol reviewed (approved):
Better to make full sync once more, in order to be sure that nothing was broken in latest changes. This can be done after merging in master, but before next release.
